src/main/java/com/google/sps/servlets/ExamServlet.java

```diff
@@ -0,0 +1,180 @@
+    response.setContentType("text/html;");
+    String examID = getParameter(request, "examID", "1");
+
+    Entity examEntity = null;
+
+    try{
+      examEntity = getEntity(examID);
+    }catch (EntityNotFoundException e){
+    }
+
+    if(examEntity != null) {
+      String name = (String) examEntity.getProperty("name");
+      String duration = (String) examEntity.getProperty("duration");
+      String ownerID = (String) examEntity.getProperty("ownerID");
+      List<Long> questionsList = (List<Long>) examEntity.getProperty("questionsList");
+      ExamClass exam = new ExamClass(name,examEntity.getKey().getId(),
+        Double.parseDouble(duration),ownerID,questionsList);
+        
+      
+      response.getWriter().println("<h1>Exam Name: " + exam.getName() + "</h1>");
+      response.getWriter().println("<h2>Length: " + exam.getDuration() + "</h2>");
+      response.getWriter().println("<h2>Created By: " + exam.getOwnerID() + "</h2>");
+      if(questionsList != null){
+        List<QuestionClass> listofQuestions=getQuestionsFromExam(questionsList);
+        response.getWriter().println(convertToJsonUsingGson(listofQuestions));
+      }else{
+        response.getWriter().println("<p>There are no questions associated with this exam.</p>");
+      }
+    } else {
+      response.getWriter().println("<h1>Choose an exam to take.</h1>");
+      response.getWriter().println("<table><tr><th>ID</th><th>Name</th><th>Duration</th></tr>");
+
+      Query query = new Query("Exam");
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+      for (Entity entity : results.asIterable()) {
+        // Send back amount of results requested
+        long id = entity.getKey().getId();
+        String name = (String) entity.getProperty("name");
+        String duration = (String) entity.getProperty("duration");
+
+      response.getWriter().println("<tr>");
+      response.getWriter().println("<td>" + id + "</td>");
+      response.getWriter().println("<td>" + name + "</td>");
+      response.getWriter().println("<td>" + duration + "</td>");
+      response.getWriter().println("<td><a href=\"/exam?examID=" + id + "\">Take Exam</a></td>");
+      response.getWriter().println("</tr>");
+      }
+      response.getWriter().println("</table>");
+    }
+
+  }
+  private List<QuestionClass> getQuestionsFromExam(List<Long> list){
+    /*Grab all the question id's from the list and creates instances of 
+    *questions from them.
+    */
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<QuestionClass> questionList = new ArrayList<>();
+    if (list == null){
+      return null;
+    }
+    for(int i=0 ; i<list.size(); i++)
+    {
+      try{
+        Key key = KeyFactory.createKey("Question", list.get(i));
+        Entity qs = datastore.get(key);
```

Comment:
I'm not familiar with datastore API, so this comment might not be applicable. However if there's a way to combine these requests into one, it will be faster, because you will reduce overhead created by communicating w/ datastore.

Reply:
Looks like it is possible to create a RPC request to get entities by key with lookup method of Google Cloud Datastore, will look more into this in next iteration!